js/plugins/vertexai/src/vector-search/bigquery.ts

```diff
@@ -0,0 +1,106 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Document, DocumentDataSchema } from '@genkit-ai/ai/retriever';
+import { BigQuery } from '@google-cloud/bigquery';
+import { DocumentIndexer, DocumentRetriever, Neighbor } from './types';
+/**
+ * Creates a BigQuery Document Retriever.
+ *
+ * This function returns a DocumentRetriever function that retrieves documents
+ * from a BigQuery table based on the provided neighbors.
+ *
+ * @param {BigQuery} bq - The BigQuery instance.
+ * @param {string} tableId - The ID of the BigQuery table.
+ * @param {string} datasetId - The ID of the BigQuery dataset.
+ * @returns {DocumentRetriever} - The DocumentRetriever function.
+ */
+export const getBigQueryDocumentRetriever = (
+  bq: BigQuery,
+  tableId: string,
+  datasetId: string
+): DocumentRetriever => {
+  const bigQueryRetriever: DocumentRetriever = async (
+    neighbors: Neighbor[]
+  ): Promise<Document[]> => {
+    const ids = neighbors
+      .map((neighbor) => neighbor.datapoint?.datapointId)
+      .filter(Boolean);
+    const query = `
+      SELECT * FROM \`${datasetId}.${tableId}\`
+      WHERE id IN UNNEST(@ids)
+    `;
+    const options = {
+      query,
+      params: { ids },
+    };
+    const [rows] = await bq.query(options);
+    const docs: (Document | null)[] = rows
+      .map((row) => {
+        const docData = {
+          content: JSON.parse(row.content),
+          metadata: {
+            ...neighbors.find(
+              (neighbor) => neighbor.datapoint?.datapointId === row.id
+            ),
+          },
+        };
+        const parsedDocData = DocumentDataSchema.safeParse(docData);
+        if (parsedDocData.success) {
+          return new Document(parsedDocData.data);
+        }
+        return null;
+      })
+      .filter((doc) => doc !== null) as Document[];
+    return docs as Document[];
+  };
+  return bigQueryRetriever;
+};
+
+/**
+ * Creates a BigQuery Document Indexer.
+ *
+ * This function returns a DocumentIndexer function that indexes documents
+ * into a BigQuery table.
+ *
+ * @param {BigQuery} bq - The BigQuery instance.
+ * @param {string} tableId - The ID of the BigQuery table.
+ * @param {string} datasetId - The ID of the BigQuery dataset.
+ * @returns {DocumentIndexer} - The DocumentIndexer function.
+ */
+export const getBigQueryDocumentIndexer = (
+  bq: BigQuery,
+  tableId: string,
+  datasetId: string
+): DocumentIndexer => {
+  const bigQueryIndexer: DocumentIndexer = async (
+    docs: Document[]
+  ): Promise<string[]> => {
+    const ids: string[] = [];
+    const rows = docs.map((doc) => {
+      const id = Math.random().toString(36).substring(7);
+      ids.push(id);
+      return {
+        id,
+        content: JSON.stringify(doc.content),
+        metadata: JSON.stringify(doc.metadata),
+      };
+    });
+    await bq.dataset(datasetId).table(tableId).insert(rows);
```

Comment:
Nit: Please make a note in the JSDoc that this indexer does not handle duplicate documents.